### PR TITLE
ADR #10: log handling

### PR DESCRIPTION
**Pull request by BrianHicks**
I've been trying to think through how we handle logs. We have a couple different cases we need to handle. I think it needs a little more structure than we have right now, so here's an ADR to talk about that.

I'd like feedback on:

- other than CLI/UI streaming, where else would it make sense to stream logs?
- are there any cases in which collecting all the logs for an invocation wouldn't make sense?

### CODE REVIEW

**Comment by BrianHicks** [comment]
@rtfeldman I intended to talk with you briefly about this during our pairing today. But, since we had to cancel, would you just mind having a look sometime this week?

**Comment by BrianHicks** [comment]
@rtfeldman not to poke *too* much but… any feedback on this approach? I'd love to get started but I don't wanna be the only one who knows what's even going on with the implementation 😆

File: docs/adrs/010-logs.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+How do we persist logs so you can look at them later, for example when a CI job fails?
+
+It seems like the best solution would be to have both a log store and a log sink.
+Specifically:
+
+1. during evaluation of jobs, logs can be streamed.
+   This would let us show them in the CLI or in a web UI.
+2. Logs could be stored like job artifacts in the content-addressable store.
+
+Part 1 can be implemented today, but part two needs a little bit more groundwork.
```
Comment (bhansconnect) [inline docs/adrs/010-logs.md:15]: If you were to do this directly, wouldn't you just get a mess of mixed logs due to concurrent jobs?

File: docs/adrs/010-logs.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+Specifically, how do you keep track of which logs came from which jobs?
+You could store them with the artifacts, but that seems unreasonable in the case of failing jobs (since it would be unreasonable to store artifacts for a failing job.)
+
+Instead, it seems like we need the concept of an invocation, which could keep track of which jobs it did and didn't build, along with log outputs.
```
Comment (bhansconnect) [inline docs/adrs/010-logs.md:19]: Invocations seem to more be answering the question of handling longs when you think across runs. It still seems that just the job hash is all that is needed to tell which logs come from which job.

File: docs/adrs/010-logs.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+You also might want to see logs as the system is building.
+This is more complicated than just "ok, so stream them" because we plan to add remote builders.
+How do we persist logs so you can look at them later, for example when a CI job fails?
+
+It seems like the best solution would be to have both a log store and a log sink.
+Specifically:
+
+1. during evaluation of jobs, logs can be streamed.
+   This would let us show them in the CLI or in a web UI.
+2. Logs could be stored like job artifacts in the content-addressable store.
```
Comment (bhansconnect) [inline docs/adrs/010-logs.md:13]: I think you want one more property. Some sort of toggle about what is display and at what verbosity.

File: docs/adrs/010-logs.md (hunk 1 above)
Comment (BrianHicks) [reply docs/adrs/010-logs.md:15]: I'm not sure what you mean, here. We wouldn't be interleaving them—they'd be stored separately, but associated with an invocation.

File: docs/adrs/010-logs.md (hunk 2 above)
Comment (BrianHicks) [reply docs/adrs/010-logs.md:19]: I'm not 100% sure what you mean here, actually. I suppose that it might be the case that logs are attached to a job, but where do we keep track of that? The invocation seems like a reasonable way to go; do you see a better place?

File: docs/adrs/010-logs.md (hunk 3 above)
Comment (BrianHicks) [reply docs/adrs/010-logs.md:13]: I think that'd probably be configured at the job level. We're just talking about stdout/stderr here. Maybe I should make that clearer up top!

**Comment by BrianHicks** [comment]
@bhansconnect reading your comments through again, I reworded this ADR to make it clearer that we're talking about job logs (e.g. stdout and stderr), not logs from rbt. Does it make more sense in that light?

File: docs/adrs/010-logs.md (hunk 1 above)
Comment (bhansconnect) [reply docs/adrs/010-logs.md:15]: I specifically meant showing stdout and stderr in the CLI. With concurrent jobs theoretically run on a cluster being streaming back to a dev machine, I feel like it would generate an unreadable mess of logs.

File: docs/adrs/010-logs.md (hunk 2 above)
Comment (bhansconnect) [reply docs/adrs/010-logs.md:19]: > Specifically, how do you keep track of which logs came from which jobs?

Currently we generate a spot to store artifacts for each job. I don't see why you couldn't just create a spot next to/with the same hash key to store the logs. If the jobs hash is `1234` and I store the artifacts in `jobs/1234/artifacts`, couldn't you just store the logs in `jobs/1234/logs`?

If the issue here is that you want to keep around logs from old job runs, an invocation hash or just timestamp sounds necessary, but I don't see how an invocation hash is really need to answer the original question.

File: docs/adrs/010-logs.md (hunk 3 above)
Comment (bhansconnect) [reply docs/adrs/010-logs.md]: makes sense. Though likely you will want some way to dynamically pipe config to jobs. So an end user can request that job 42 be more verbose by only adding a command line flag to `rbt` rather than needing to update the job level config in the `.roc` file.

**Review by bhansconnect — APPROVED** [review]
The ADR totally makes sense. This comments are overall relatively small. I tried to clarify them better.

File: docs/adrs/010-logs.md (hunk 2 above)
Comment (BrianHicks) [reply docs/adrs/010-logs.md:19]: > I don't see why you couldn't just create a spot next to/with the same hash key to store the logs.

Two reasons:

1. we store artifacts in a content-addressable store. Logs frequently have timestamps in them. That's good! But it also means that then our artifacts wouldn't be reproducible.
  - this means I'm a little leery of putting them in the CAS *at all* but it's also fine? I expect we'll have to make some sort of garbage collector for the CAS at some point regardless of this decision. If anything, this'll just speed that up.
3. it wouldn't always be reasonable to store artifacts for a failed build (I'm not sure if it'd ever be, actually?) But it'd be always-reasonable to store logs for a failed build, for debugging after the fact.

Basically: they do different jobs and have slightly different storage needs.
